### PR TITLE
added browser back functionality

### PR DESCRIPTION
**Pull request by WMDE-Fisch**
Don't judge my code - I was young and needed this job!

### CODE REVIEW

**Comment by tobijat** [comment]
:)))

File: app/Questionnaire.js [hunk 1]
```diff
@@ -98,10 +104,33 @@ $.extend( Questionnaire.prototype, {
 	 *         - {AjaxError}
 	 */
 	start: function() {
+                var self = this;
```
Comment (thiemowmde) [inline app/Questionnaire.js]: Looks like this code has indention issues. Please use tabs everywhere.

File: app/Questionnaire.js [hunk 2]
```diff
@@ -98,10 +104,33 @@ $.extend( Questionnaire.prototype, {
...
 
 		this._navigationCache = [];
+                this._navigationState = 0;
+                
+                //cf. set initial history state
+                History.pushState( { nr: this._navigationState }, '', '?state=0' );
+            
+                //cf. add event listener for history state changes
+                History.Adapter.bind( window, 'statechange', function() {
+                        var State = History.getState(),
```
Comment (thiemowmde) [inline app/Questionnaire.js]: A simple variable shouldn't be upper case.

File: app/Questionnaire.js [hunk 3]
```diff
@@ -477,21 +507,47 @@ $.extend( Questionnaire.prototype, {
...
-			if( self._navigationCache.length === 0 ) {
-				$( self ).trigger( 'back', [self._asset] );
-			} else {
-				self._goTo( self._navigationCache[self._navigationCache.length - 1].page, true )
-				.done( function() {
-					$( self ).trigger( 'update' );
-				} );
-			}
+		$backButton.on( 'click', function() {			
+                        History.back();
```
Comment (thiemowmde) [inline app/Questionnaire.js]: Hurray! But, ... hm, isn't this the default anyway?

File: app/Questionnaire.js [hunk 4]
```diff
@@ -477,21 +507,47 @@ $.extend( Questionnaire.prototype, {
...
 		return $backButton;
 	},
 
 	/**
+	 * Navigates one step forward using the cache.
+	 * christoph.fischer
+	 */
+	_navigateForward: function() {
+		var self = this;
+		self._navigationState++;
```
Comment (thiemowmde) [inline app/Questionnaire.js]: Stick to `this` outside of the closure that requires `self`.

File: app/Questionnaire.js [hunk 5]
```diff
@@ -1112,6 +1165,38 @@ $.extend( Questionnaire.prototype, {
...
+                            loggedAnswers: ({}, {})
+                    };
+
+
+                this._navigationCache.splice(  this._navigationState );
+
+                if( this._navigationCache.length === 0 ) {
+                    this._navigationCache.push( updatedNavigationEntry );
+                } else {
+                    this._navigationCache[this._navigationState] = updatedNavigationEntry;
```
Comment (thiemowmde) [inline app/Questionnaire.js]: This is odd. The `push` above adds the first element with **no** key, but later elements are added **with** a key?

File: app/Questionnaire.js [hunk 6]
```diff
@@ -1112,6 +1165,38 @@ $.extend( Questionnaire.prototype, {
...
+                }
+
+                this._navigationCache.push( nextNavigationEntry );
+
+                this._navigationState++;
```
Comment (thiemowmde) [inline app/Questionnaire.js]: I wonder if this code would work if all instances of `_navigationState` are replaced with `_navigationCache.length`?

**Comment by wmde-manicki** [comment]
Thank you @thiemowmde for the thorough review. The code posted by @WMDE-Fisch is not intended to be merged as such. We will develop some other solution based on what has been done on this code. I have created some initial alternative solution in the `history-back` branch and would make a pull request once I am happy with that (there are some things missing still).

**Comment by thiemowmde** [comment]
Ok, thanks for letting me know!

**Comment by WMDE-Fisch** [comment]
I will close this PR since we implemented an other solution #70 . We might leave the branch though.

**Pull request closed without merging**